### PR TITLE
Cleanup ServerLoader logic

### PR DESCRIPTION
**Pull request by HaoK**
Fixes https://github.com/aspnet/Hosting/issues/214

Note looks like the old code was doing a weird (finding all IServerLoaders and the last one wins).  Now we just select the first one (I'm not sure that edge case matters, but we could throw possibly instead and use SingleOrDefault?)

cc @Tratcher @davidfowl

### CODE REVIEW

File: src/Microsoft.AspNet.Hosting/Server/ServerLoader.cs [hunk 1]
```diff
@@ -18,59 +18,29 @@ public ServerLoader(IServiceProvider services)
...
-                foreach (var typeInfo in assembly.DefinedTypes)
-                {
-                    interfaceInfo = typeInfo.ImplementedInterfaces.FirstOrDefault(interf => interf.Equals(typeof(IServerFactory)));
-                    if (interfaceInfo != null)
-                    {
-                        type = typeInfo.AsType();
-                    }
-                }
+            var serverType = assembly.DefinedTypes.Where(
+                t => t.ImplementedInterfaces.FirstOrDefault(interf => interf.Equals(typeof(IServerFactory))) != null)
```
Comment (hishamco) [inline src/Microsoft.AspNet.Hosting/Server/ServerLoader.cs]: Can't convert ```TypeInfo``` to ```System.Type```

Comment (HaoK) [inline src/Microsoft.AspNet.Hosting/Server/ServerLoader.cs]: Ah thanks, I don't think we use this entry point very often, I'm not sure we have any unit test coverage at all currently of this...  I'll see if I can find someplace to add one for this

Comment (HaoK) [inline src/Microsoft.AspNet.Hosting/Server/ServerLoader.cs]: Actually TypeInfo is a Type, so I'm not sure AsType is actually needed

Comment (hishamco) [inline src/Microsoft.AspNet.Hosting/Server/ServerLoader.cs]: Or you can modify the query like this
```csharp
            var serverType = assembly.DefinedTypes.Where(ti=>ti.ImplementedInterfaces.Contains(typeof(IServerFactory))).Select(t=>t.AsType()).FirstOrDefault();
```

File: src/Microsoft.AspNet.Hosting/Server/ServerLoader.cs [hunk 2]
```diff
@@ -18,59 +18,29 @@ public ServerLoader(IServiceProvider services)
...
         }
 
-        public IServerFactory LoadServerFactory(string serverFactoryIdentifier)
+        public IServerFactory LoadServerFactory([NotNull] string assemblyName)
         {
-            if (string.IsNullOrEmpty(serverFactoryIdentifier))
+            if (string.IsNullOrEmpty(assemblyName))
             {
-                throw new ArgumentException(string.Empty, nameof(serverFactoryIdentifier));
+                throw new ArgumentException(string.Empty, nameof(assemblyName));
```
Comment (henkmollema) [inline src/Microsoft.AspNet.Hosting/Server/ServerLoader.cs]: Do we still need this or does `[NotNull]` not check for empty strings?

Comment (HaoK) [inline src/Microsoft.AspNet.Hosting/Server/ServerLoader.cs]: Both, NotNull doesn't yet enforce anything at runtime, and no it won't prevent empty strings

File: src/Microsoft.AspNet.Hosting/Server/ServerLoader.cs (hunk 1 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Hosting/Server/ServerLoader.cs]: Looks like the ServerTests repo exercise this code path today (WebListener as well)

**Comment by HaoK** [comment]
@Tratcher can you review?

File: src/Microsoft.AspNet.Hosting/Server/ServerLoader.cs [hunk 3]
```diff
@@ -18,59 +18,29 @@ public ServerLoader(IServiceProvider services)
             _services = services;
         }
 
-        public IServerFactory LoadServerFactory(string serverFactoryIdentifier)
+        public IServerFactory LoadServerFactory([NotNull] string assemblyName)
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Hosting/Server/ServerLoader.cs]: Methods implementing interfaces should use the same parameter names as the interface. `serverName` in this case.

**Comment by Tratcher** [comment]
:shipit:

**Comment by HaoK** [comment]
bf6e293bfedcb921215f8c4273839912be205719

**Pull request closed without merging**